flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/PartitionSpecEvolver.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.dynamic;
+
+import java.util.List;
+import java.util.Objects;
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.expressions.NamedReference;
+import org.apache.iceberg.expressions.Term;
+import org.apache.iceberg.expressions.UnboundTransform;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/** Checks compatibility of PartitionSpecs and evolves one into the other. */
+public class PartitionSpecEvolver {
+
+  private PartitionSpecEvolver() {}
+
+  /**
+   * Checks whether two PartitionSpecs are compatible with each other. Less strict than {@code
```

Comment:
Can you share why a different/relaxed compatibility check is needed? what are the scenarios?

Reply:
The default PartitionSpec.equals() compares the specs based on the transform name and source ids. Since we don't enforce source ids to match in the Dynamic Sink, but check the actual field names, we need the custom logic.